File: app/java/src/main/java/com/example/production/application/service/CompletionResultService.java
package com.example.production.application.service;

import com.example.production.application.port.in.command.CompletionResultCreateCommand;
import com.example.production.application.port.out.CompletionInspectionResultRepository;
import com.example.production.application.port.out.CompletionResultRepository;
import com.example.production.application.port.out.WorkOrderRepository;
import com.example.production.domain.model.process.CompletionInspectionResult;
import com.example.production.domain.model.process.CompletionResult;
import com.example.production.domain.model.process.WorkOrder;
import com.example.production.domain.model.process.WorkOrderStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * 完成実績サービス
 */
@Service
@RequiredArgsConstructor
public class CompletionResultService {

    private final CompletionResultRepository completionResultRepository;
    private final CompletionInspectionResultRepository completionInspectionResultRepository;
    private final WorkOrderRepository workOrderRepository;

    /**
     * 完成実績番号を生成する
     */
    private String generateCompletionResultNumber(LocalDate completionDate) {
        String prefix = "CR-" + completionDate.format(DateTimeFormatter.ofPattern("yyyyMM")) + "-";
        return completionResultRepository.findLatestCompletionResultNumber(prefix)
                .map(latest -> {
                    int currentSequence = Integer.parseInt(latest.substring(latest.length() - 4));
                    return prefix + String.format("%04d", currentSequence + 1);
                })
                .orElse(prefix + "0001");
    }

    /**
     * 完成実績を登録する
     */
    @Transactional
    public CompletionResult createCompletionResult(CompletionResultCreateCommand command) {
        // 作業指示を取得
        WorkOrder workOrder = workOrderRepository.findByWorkOrderNumber(command.getWorkOrderNumber())
                .orElseThrow(() -> new IllegalArgumentException("Work order not found: " + command.getWorkOrderNumber()));

        // 作業指示の状態を確認
        if (workOrder.getStatus() != WorkOrderStatus.IN_PROGRESS) {
            throw new IllegalStateException("Only IN_PROGRESS work orders can record completion results");
        }

        String completionResultNumber = generateCompletionResultNumber(command.getCompletionDate());

        // 完成実績を作成
        CompletionResult completionResult = CompletionResult.builder()
                .completionResultNumber(completionResultNumber)
                .workOrderNumber(command.getWorkOrderNumber())
                .itemCode(workOrder.getItemCode())
                .completionDate(command.getCompletionDate())
                .completedQuantity(command.getCompletedQuantity())
                .goodQuantity(command.getGoodQuantity())
                .defectQuantity(command.getDefectQuantity())
                .remarks(command.getRemarks())
                .createdBy(command.getCreatedBy())
                .updatedBy(command.getCreatedBy())
                .build();
        completionResultRepository.save(completionResult);

        // 検査結果を登録
        List<CompletionInspectionResult> inspectionResults = new ArrayList<>();
        if (command.getInspectionResults() != null) {
            for (CompletionResultCreateCommand.InspectionResultCommand inspCmd : command.getInspectionResults()) {
                CompletionInspectionResult inspectionResult = CompletionInspectionResult.builder()
                        .completionResultNumber(completionResultNumber)
                        .defectCode(inspCmd.getDefectCode())
                        .quantity(inspCmd.getQuantity())
                        .build();
                completionInspectionResultRepository.save(inspectionResult);
                inspectionResults.add(inspectionResult);
            }
        }

        // 作業指示の完成数量を更新
        workOrderRepository.updateCompletionQuantities(
                command.getWorkOrderNumber(),
                workOrder.getCompletedQuantity().add(command.getCompletedQuantity()),
                workOrder.getTotalGoodQuantity().add(command.getGoodQuantity()),
                workOrder.getTotalDefectQuantity().add(command.getDefectQuantity())
        );

        completionResult.setInspectionResults(inspectionResults);
        return completionResult;
    }

    /**
     * 完成実績番号で検索する
     */
    public CompletionResult findByCompletionResultNumber(String completionResultNumber) {
        CompletionResult result = completionResultRepository.findByCompletionResultNumber(completionResultNumber)
                .orElse(null);
        if (result != null) {
            result.setInspectionResults(completionInspectionResultRepository.findByCompletionResultNumber(completionResultNumber));
        }
        return result;
    }

    /**
     * 作業指示番号で完成実績を検索する
     */
    public List<CompletionResult> findByWorkOrderNumber(String workOrderNumber) {
        List<CompletionResult> results = completionResultRepository.findByWorkOrderNumber(workOrderNumber);
        for (CompletionResult result : results) {
            result.setInspectionResults(completionInspectionResultRepository.findByCompletionResultNumber(result.getCompletionResultNumber()));
        }
        return results;
    }
}

File: app/csharp/src/ProductionManagement.Application/Services/CompletionResultService.cs
using ProductionManagement.Application.Port.In.Command;
using ProductionManagement.Application.Port.Out;
using ProductionManagement.Domain.Models.Process;

namespace ProductionManagement.Application.Services;

/// <summary>
/// 完成実績サービス
/// </summary>
public class CompletionResultService
{
    private readonly ICompletionResultRepository _completionResultRepository;
    private readonly ICompletionInspectionResultRepository _completionInspectionResultRepository;
    private readonly IWorkOrderRepository _workOrderRepository;

    public CompletionResultService(
        ICompletionResultRepository completionResultRepository,
        ICompletionInspectionResultRepository completionInspectionResultRepository,
        IWorkOrderRepository workOrderRepository)
    {
        _completionResultRepository = completionResultRepository;
        _completionInspectionResultRepository = completionInspectionResultRepository;
        _workOrderRepository = workOrderRepository;
    }

    /// <summary>
    /// 完成実績番号を生成する
    /// </summary>
    private async Task<string> GenerateCompletionResultNumberAsync(DateOnly completionDate)
    {
        var prefix = $"CR-{completionDate:yyyyMM}-";
        var latestNumber = await _completionResultRepository.FindLatestCompletionResultNumberAsync($"{prefix}%");

        var sequence = 1;
        if (latestNumber != null)
        {
            var currentSequence = int.Parse(latestNumber[^4..]);
            sequence = currentSequence + 1;
        }

        return $"{prefix}{sequence:D4}";
    }

    /// <summary>
    /// 完成実績を報告する
    /// </summary>
    public async Task<CompletionResult> ReportCompletionAsync(CompletionResultCommand command)
    {
        // 作業指示を取得
        var workOrder = await _workOrderRepository.FindByWorkOrderNumberAsync(command.WorkOrderNumber)
            ?? throw new ArgumentException($"Work order not found: {command.WorkOrderNumber}");

        if (workOrder.Status != WorkOrderStatus.InProgress)
        {
            throw new InvalidOperationException("Only IN_PROGRESS work orders can report completion");
        }

        var completionResultNumber = await GenerateCompletionResultNumberAsync(command.CompletionDate);

        // 完成実績を作成
        var completionResult = new CompletionResult
        {
            CompletionResultNumber = completionResultNumber,
            WorkOrderNumber = command.WorkOrderNumber,
            ItemCode = workOrder.ItemCode,
            CompletionDate = command.CompletionDate,
            CompletedQuantity = command.CompletedQuantity,
            GoodQuantity = command.GoodQuantity,
            DefectQuantity = command.DefectQuantity,
            Remarks = command.Remarks
        };
        await _completionResultRepository.SaveAsync(completionResult);

        // 検査結果を作成
        var inspectionResults = new List<CompletionInspectionResult>();
        if (command.InspectionResults != null)
        {
            foreach (var irCommand in command.InspectionResults)
            {
                var ir = new CompletionInspectionResult
                {
                    CompletionResultNumber = completionResultNumber,
                    DefectCode = irCommand.DefectCode,
                    Quantity = irCommand.Quantity
                };
                await _completionInspectionResultRepository.SaveAsync(ir);
                inspectionResults.Add(ir);
            }
        }

        // 作業指示の累計を更新
        await _workOrderRepository.UpdateCompletionQuantitiesAsync(
            command.WorkOrderNumber,
            command.CompletedQuantity,
            command.GoodQuantity,
            command.DefectQuantity
        );

        completionResult.InspectionResults = inspectionResults;
        return completionResult;
    }

    /// <summary>
    /// 完成実績を検索する
    /// </summary>
    public async Task<CompletionResult?> FindByCompletionResultNumberAsync(string completionResultNumber)
    {
        var completionResult = await _completionResultRepository.FindByCompletionResultNumberAsync(completionResultNumber);
        if (completionResult != null)
        {
            var inspectionResults = await _completionInspectionResultRepository.FindByCompletionResultNumberAsync(completionResultNumber);
            completionResult.InspectionResults = inspectionResults;
        }
        return completionResult;
    }

    /// <summary>
    /// 作業指示に紐づく完成実績を検索する
    /// </summary>
    public async Task<IReadOnlyList<CompletionResult>> FindByWorkOrderNumberAsync(string workOrderNumber)
    {
        return await _completionResultRepository.FindByWorkOrderNumberAsync(workOrderNumber);
    }
}
